File: bak/imageStore.cpp
#include "bak/imageStore.hpp"
#include "bak/dataTags.hpp"

#include "com/logger.hpp"

#include <iostream>

namespace BAK {

// create a logger for this file.
const auto logger = Logging::LogState::GetLogger("imageStore");

std::vector<Image> LoadImagesNormal(FileBuffer& fb)
{
    std::vector<Image> images{};

    const unsigned compression = fb.GetUint16LE();
    logger.Info() << "compression: " << compression << std::endl;
    const unsigned numImages = fb.GetUint16LE();
    logger.Info() << "numImages: " << numImages << std::endl;
    std::vector<unsigned> imageSizes{};
    fb.Skip(2);
    unsigned int size = fb.GetUint32LE();
    logger.Info() << "size: " << size << std::endl;
    for (unsigned i = 0; i < numImages; i++)
    {
        const auto imageSize = fb.GetUint16LE();
        logger.Info() << "imageSize: " << imageSize << std::endl;
        imageSizes.emplace_back(imageSize);
        unsigned flags = fb.GetUint16LE();
        logger.Info() << "flags: " << flags << std::endl;
        unsigned width = fb.GetUint16LE();
        logger.Info() << "width: " << width << std::endl;
        unsigned height = fb.GetUint16LE();
        logger.Info() << "height: " << height << std::endl;
        images.emplace_back(width, height, flags, false);
    }

    if (compression == 1)
    {
        // Not sure why this is needed or if *2 is the right number
        size *= 2;
    }

    FileBuffer decompressed = FileBuffer(size);

    // Log the fb buffer
    // const uint8_t* buffer = fb.GetBuffer();
    // const size_t bufferSize = fb.GetSize();
    // std::cout << "fb.GetBuffer(): [";
    // for (size_t i = 0; i < bufferSize; ++i)
    // {
    //     std::cout << static_cast<int>(buffer[i]);
    //     if (i < bufferSize - 1)
    //         std::cout << ", ";
    // }
    // std::cout << "]" << std::endl;

    logger.Info() << "before decompress: decompressed.length: " << decompressed.GetSize() << std::endl;
    logger.Info() << "before decompress: fb.length: " << fb.GetSize() << std::endl;

    fb.Decompress(&decompressed, compression);
    for (unsigned int i = 0; i < numImages; i++)
    {
        auto imageBuffer = FileBuffer(imageSizes[i]);
        imageBuffer.Fill(&decompressed);

        {
            // Log the imageBuffer buffer
            const uint8_t* buffer = imageBuffer.GetBuffer();
            const size_t bufferSize = imageBuffer.GetSize();
            std::cout << "imageBuffer.GetBuffer(): [";
            for (size_t i = 0; i < bufferSize; ++i)
            {
                std::cout << static_cast<int>(buffer[i]);
                if (i < bufferSize - 1)
                    std::cout << ", ";
            }
            std::cout << "]" << std::endl;
        }

        images[i].Load(&imageBuffer);

        {
            // Log the images[i].pixels
            const uint8_t* pixels = images[i].GetPixels();
            const size_t pixelsSize = images[i].GetSize();
            std::cout << "images[" << i << "].pixels: [";
            for (size_t j = 0; j < pixelsSize; ++j)
            {
                std::cout << static_cast<int>(pixels[j]);
                if (j < pixelsSize - 1)
                    std::cout << ", ";
            }
        }
    }

    return images;
}

std::vector<Image> LoadImagesTagged(FileBuffer& fb)
{
    auto infBuf = fb.Find(DataTag::INF);

    std::vector<Image> images{};
    unsigned imageCount = infBuf.GetUint16LE();
    for (unsigned i = 0; i < imageCount; i++)
    {
        const auto width = infBuf.GetUint16LE();
        const auto start = infBuf.Tell();
        infBuf.Skip(2 * (imageCount - 1));
        const auto height = infBuf.GetUint16LE();
        infBuf.Seek(start);
        images.emplace_back(width, height, 0, true);
    }

    auto binBuf = fb.Find(DataTag::BIN);
    auto compression = binBuf.GetUint8();
    auto size = binBuf.GetUint32LE();
    FileBuffer decompressed = FileBuffer(size);
    auto decompressedBytes = binBuf.DecompressLZW(&decompressed);
    for (unsigned i = 0; i < imageCount; i++)
    {
        images[i].Load(&decompressed);
    }

    return images;
}

std::vector<Image> LoadImages(FileBuffer& fb)
{
    const auto imageType = fb.GetUint16LE();
    if (imageType == 0x1066)
    {
        return LoadImagesNormal(fb);
    }
    else if (imageType == 0x4d42)
    {
        return LoadImagesTagged(fb);
    }
    else
    {
        throw std::runtime_error("Couldn't load images");
        return std::vector<Image>{};
    }
}

}

File: ts/bak/imageStore.ts
import { FileBuffer } from "./file/fileBuffer.ts";
import { extractedDataPath } from "../consts.ts";
import { COMPRESSION_LZSS } from "./file/fileBuffer.ts";
import { Image } from "./image.ts";
import { DataTag } from "./dataTags.ts";

function log(message: string) {
    console.log(`[imageStore.ts/loadImages] ${message}`);
}

// loads images from a *.BMX file. (maybe from other files???)
export function loadImagesNormal(fb: FileBuffer): Image[] {
    const images: Image[] = [];

    const compression = fb.getUint16LE();
    log(`Compression: ${compression}`);
    const numImages = fb.getUint16LE();
    log(`Number of images: ${numImages}`);

    // TODO: What is imageSizes?
    const imageSizes: number[] = [];
    fb.skip(2);
    let size = fb.getUint32LE();
    log(`Size: ${size}`);
    for (let i = 0; i < numImages; i++) {
        const imageSize = fb.getUint16LE();
        log(`Image size: ${imageSize}`);
        imageSizes.push(imageSize);
        const flags = fb.getUint16LE();
        log(`Flags: ${flags}`);
        const width = fb.getUint16LE();
        log(`Width: ${width}`);
        const height = fb.getUint16LE();
        log(`Height: ${height}`);
        images.push(new Image({
            width,
            height,
            flags,
            isHighResLowCol: false,
        }));
    }

    // log(`imageSizes: ${JSON.stringify(imageSizes)}`);
    // log(`images: ${JSON.stringify(images, null, 2)}`);

    // if (compression === 1) {
    if (compression === COMPRESSION_LZSS) {
        // Not sure why this is needed or if *2 is the right number
        // TODO: Examine the bahaviour. should be easy to see how much
        // space is needed if I decompress every single image I think.
        size *= 2;
    }

    // FileBuffer decompressed = FileBuffer(size);
    // fb.Decompress(&decompressed, compression);
    // for (unsigned int i = 0; i < numImages; i++)
    // {
    //     auto imageBuffer = FileBuffer(imageSizes[i]);
    //     imageBuffer.Fill(&decompressed);
    //     images[i].Load(&imageBuffer);
    // }

    const decompressedFileBuffer = FileBuffer.createEmpty(size);
    // console.log(`fb.uint8Array: ${fb.uint8Array}`);
    // console.log(`decompressedFileBuffer.uint8Array: ${decompressedFileBuffer.uint8Array}`);
    console.log(`before decompress: decompressedFileBuffer.length: ${decompressedFileBuffer.uint8Array.length}`);
    console.log(`before decompress: fb.length: ${fb.uint8Array.length}`);
    fb.decompress(decompressedFileBuffer, compression);
    for (let i = 0; i < numImages; i++) {
        const imageBuffer = FileBuffer.createEmpty(imageSizes[i]);
        imageBuffer.fill(decompressedFileBuffer);
        // console.log(`imageBuffer.uint8Array: ${imageBuffer.uint8Array}`);
        // console.log();
        images[i].load(imageBuffer);
        // console.log(`images[${i}].pixels: ${images[i].pixels}`);
        // TODO: Seems everything is working up to this point.
    }

    return images;
}

// loads images from a *.BMX file.
export async function loadImages(resourceFile: string): Promise<Image[]> {
    log(`Loading image ${resourceFile}`);
    const path = `${extractedDataPath}/${resourceFile}`;
    log(`Loading images from ${path}`);
    const fb = await FileBuffer.createFromFile(path);

    const imageType = fb.getUint16LE();
    log(`Image type: ${imageType}`);

    if (imageType === 0x1066) {
        return loadImagesNormal(fb);
    } else if (imageType === 0x4d42) {
        return loadImagesTagged(fb);
    } else {
        throw new Error(`Couldn't load images`);
    }
}

export function loadImagesTagged(fb: FileBuffer): Image[] {
    const infBuf = fb.find(DataTag.INF);
    const images: Image[] = [];
    
    const imageCount = infBuf.getUint16LE();
    console.log(`loadImagesTagged:Image count: ${imageCount}`);
    
    for (let i = 0; i < imageCount; i++) {
        const width = infBuf.getUint16LE();
        const start = infBuf.index;
        infBuf.skip(2 * (imageCount - 1));
        const height = infBuf.getUint16LE();
        infBuf.jumpToIndex(start);
        
        images.push(new Image({
            width,
            height,
            flags: 0,
            isHighResLowCol: true
        }));
    }

    const binBuf = fb.find(DataTag.BIN);
    const compression = binBuf.getUint8();
    const size = binBuf.getUint32LE();
    
    const decompressed = FileBuffer.createEmpty(size);
    binBuf.decompressLZW(decompressed);
    
    for (let i = 0; i < imageCount; i++) {
        images[i].load(decompressed);
    }

    return images;
}

